File: src/initial_pose.cpp
#include <rclcpp/rclcpp.hpp>
#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
#include  <thread>
#include <chrono>

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    
    auto node = rclcpp::Node::make_shared("initial_pose_node");
    auto publisher = node->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>("/initialpose", 10);

    RCLCPP_INFO(node->get_logger(), "Publishing initial pose...");
    std::this_thread::sleep_for(std::chrono::seconds(3));

    auto msg = geometry_msgs::msg::PoseWithCovarianceStamped();

    msg.header.stamp = node->get_clock()->now();
    msg.header.frame_id = "map";
    msg.pose.pose.position.x = -5.0;
    msg.pose.pose.position.y = 0.0;
    msg.pose.pose.position.z = 0.0;
    msg.pose.pose.orientation.x = 0.0;
    msg.pose.pose.orientation.y = 0.0;
    msg.pose.pose.orientation.z = 0.0;
    msg.pose.pose.orientation.w = 1.0;
    msg.pose.covariance = {
        0.25, 0.0, 0.0, 0.0, 0.0, 0.0,
        0.0, 0.25, 0.0, 0.0, 0.0, 0.0,
        0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
        0.0, 0.0, 0.0, 0.0685, 0.0, 0.0,
        0.0, 0.0, 0.0, 0.0, 0.0685, 0.0,
        0.0, 0.0, 0.0, 0.0, 0.0, 0.0685
    };

    publisher->publish(msg);
    RCLCPP_INFO(node->get_logger(), "Initial pose published. Node shutting down.");
    rclcpp::shutdown();
    return 0;
}

File: turtlebot_test/initial_pose.py
import rclpy
import math
from rclpy.node import Node

from geometry_msgs.msg import PoseWithCovarianceStamped

class InitialPose(Node):
    def __init__(self):
        super().__init__("initial_pose_node")
        self.publisher_ = self.create_publisher(PoseWithCovarianceStamped, "/initialpose", 10)

        self.declare_parameter("x", 0.0)
        self.declare_parameter("y", 0.0)
        self.declare_parameter("yaw", 0.0)

        x = self.get_parameter("x").get_parameter_value().double_value
        y = self.get_parameter("y").get_parameter_value().double_value
        yaw = self.get_parameter("yaw").get_parameter_value().double_value

        msg = PoseWithCovarianceStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = "map"
        msg.pose.pose.position.x = x
        msg.pose.pose.position.y = y
        msg.pose.pose.position.z = 0.0
        msg.pose.pose.orientation.x = 0.0
        msg.pose.pose.orientation.y = 0.0
        msg.pose.pose.orientation.z = math.sin(yaw / 2.0)
        msg.pose.pose.orientation.w = math.cos(yaw / 2.0)

        msg.pose.covariance = [
            0.25, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.25, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0685, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0685, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0685
        ]

        self.publisher_.publish(msg)
        self.get_logger().info("Initial pose published.")


def main(args=None):
    rclpy.init(args=args)
    rclpy.spin(InitialPose())
    rclpy.shutdown()

if __name__ == '__main__':
    main()